docs/rfc/rfc-026-p2p-bad-peers-checktx.md

```diff
@@ -0,0 +1,375 @@
+
+For b) we need to understand what is the potential reason a transaction could never have been valid on one node, but passes `CheckTx` on another node.  
+We need to understand all the possible scenarios in which this can happen:
+
+1. What happens if a node is misconfigured and allows, for example, very large transactions into the mempool. 
+This node would then gossip these transactions and they would always fail on other nodes. 
+Is this a scenario where we want nodes to disconnect from this peer and ban it without this peer being neccessarily malicious?
+2. Are all other reasons for this to happen sign of malicious behaviour where a node explicitly lies? How can `CheckTx` pass on a valid node, but fail on another valid node with a `ResponseCheckTx.code > 1`?
+If such behaviour is only possible when a peer is malicious, should this peer be punished or banned forever? Note that 
+we cannot know whether a node is a validator in order for it to be punished. Gossiping this behaviour to other peers pro-actively
+also entails a different set of problems with it - how do we know we can trust peers who tell us to ban other peers. For these reasons, understanding the actual reason for these failures can be left for future work. 
+
+For now, we will disconnect and ban the peer regardless of the exact reason a transaction 
+is considered to never be valid. 
+
+#### **Banning in case of ResponseCheckTx.code = 1**
+
+If a node sends transactions that fail `CheckTx` with a response code of 1, there might be a valid reason for this behaviour. 
+A peer should not be banned the first time this happens, but rather if this happens a predefined number 
+of times (`numFailures`) within a time interval (`lastFailure`). This time interval should be reset every `failureResetInterval`ms.  
+
+For each peer, we should have a separate `numFailures` and `lastFailure` variable. There is no need to  have one per transaction. Whenever a transaction fails, if the `now - lastFailure <= failureResetInterval`, then we increment the `numFailures` for this particular peer and set the `lastFailure` to `now`. Otherwise, we set `lastFailure` to `now` but do not increment the `numFailures`  variable. Once the value for `numFailures` for a peer reaches `maxAllowedFailures`, the peer is disconnected from and banned. 
+
+The reason for this logic is as follows: We deem it acceptable if every now and then a peer sends us an invalid transaction. But if this happens very frequently, then this behaviour can be considered as spamming and we want to disconnect from the peer. The actual values of these parameter are yet to be defined and should be based on the time it usually takes for a transacation to travel between two nodes, the time for a block to be agreed on and committed. [ToDo] What other factors can impact their value?
```

Comment:
For this case, if we are unable to come up with a convincing property or invariant at spec level, I would advocate for banning only temporarily, as this might the due to timing, or network asynchrony glitches, which nodes have no way of eliminating.

Reply:
I agree.. The current p2p implementation however, does not have that option.